Add external secrets integration doc (#1738)
ngrok/ngrok-docs: integrations/kubernetes-ingress/external-secrets.mdx

---
title: ngrok External Secrets integration
description: ngrok integrates with External Secrets Operator to enable secure management of secrets used in ngrok Traffic Policy.
sidebarTitle: External Secrets
---

This integration allows you to push Kubernetes secrets to ngrok vaults, where they can be referenced in your traffic policies for secure configuration management.

## How it works

[External Secrets Operator](https://external-secrets.io/) watches for PushSecret resources in your cluster.
When a PushSecret is created or updated, it reads the specified Kubernetes secret and pushes the secret data to your ngrok vault using the ngrok API.
The secret then becomes available in ngrok for use in Traffic Policies.
The operator continues to sync changes based on the configured refresh interval, ensuring your ngrok secrets stay up-to-date with changes in Kubernetes.

### Limitations

> - Currently supports push operations only (Kubernetes → ngrok)

Reviewer: ```suggestion
- Currently supports push operations only (External Secret Store → ngrok)
```

@jonstacks Do we want to say Kubernetes → ngrok? Or something more... generic?

Reviewer 2: I think its fine. Maybe `Kubernetes Secrets` if we are really wanting to change it.

- Pull operations (ngrok → Kubernetes) are not yet supported

## Use cases

> - **Traffic Policy configuration:** Store API keys, tokens, and credentials used in ngrok Traffic Policy rules

Reviewer 3: ```suggestion
- **Traffic Policy configuration:** Store API keys, tokens, and credentials that you can use to authenticate visitors with traffic policy actions
```
Only gotta capitalize the first "traffic policy" under the H2 (maybe this rule is more confusing than we initially thought 😅. I may have forgotten to correct the capitalization in a few places). 

I think we mean tp actions here, not rules?

Also, I added a little bit about auth to make the actual value add more concrete here

- **Multi-environment management:** Sync different secrets to different ngrok vaults for dev, staging, and production
- **Secret rotation:** Automatically propagate rotated secrets from Kubernetes to ngrok

> - **Compliance:** Keep sensitive configuration data out of Traffic Policy definitions

Reviewer 3: ```suggestion
- **Compliance:** Use sensitive configuration data in traffic policies securely, without leaking them
```
Again, just trying to be a little more explicit about the value add

## What you'll need

- ngrok account with API access
- External Secrets Operator installed in your Kubernetes cluster
- ngrok API key

> - An ngrok vault created for storing secrets

Reviewer 3: ```suggestion
- An ngrok account with API access
- An ngrok API key
- An ngrok vault created for storing secrets
- External Secrets Operator installed in your Kubernetes cluster
```
Just added "An" and reordered things so all the ngrok stuff is grouped

spec:
  deletionPolicy: Delete
  refreshInterval: 10m  # Refresh interval for which push secret will reconcile
  secretStoreRefs:  # A list of secret stores to push secrets to
    - name: ngrok  # Must match SecretStore on the cluster
      kind: SecretStore
  selector:
    secret:
      name: SECRET_NAME  # Source Kubernetes secret to be pushed
  data:
    - match:
        # The key in the Kubernetes secret to push. Leave empty to push all keys, JSON encoded.
        # secretKey: ""
        secretKey: MY_K8S_SECRET_KEY
        remoteRef:
          remoteKey: MY_NGROK_SECRET_NAME  # The name of the secret in the ngrok vault
      metadata:
        apiVersion: kubernetes.external-secrets.io/v1alpha1
        kind: PushSecretMetadata
        spec:
          # See https://ngrok.com/docs/api/resources/secrets/#parameters
          # We currently support customizing the description and metadata for the secret.
          description: "This is a secret for the API credentials"
          # Metadata for the secret in the ngrok vault. This will be merged with auto-generated metadata.
          metadata:
            environment: production
            team: devops

> ```

Author: it does end pretty abruptly. We could add links to other more general resources to stuff about the k8s operator, secrets, traffic policy etc if you think it would be useful
